Use httptest to provide a splunk-free way of running tests

antifuchs opened the pull request:
This PR updates the tests to not require running a splunk instance (I don't know how to configure one, but I can read the API docs and provide a fake response!). 

This way, we can run the tests on a stand-alone box/container and even run them on CI systems (:

antifuchs committed:
Use httptest to provide a splunk-free way of running tests

This stubs out the response in a pretty simple-minded way, but it
probably beats having to run splunk (lite? configured how?) locally in
order to run tests.

fuyufjh commented:
Thanks @antifuchs 

The mock http server always returns `Success`. This seems not strict enough, for example even if an error happened and the client sends a totally broken JSON string, it will still give `Success` response.

I think the mock server may do some necessary check, such as parse the JSON string and checks necessary field, OR just expecting a given string for each test case.

antifuchs commented:
It's true - the mock server is quite permissive. I did this intentionally, since the tests were exercising the happy path. Tests for unhappy responses would be great & shouldn't be too hard to construct; I'll make one as an example.

antifuchs committed:
Test error handling

This test simulates a server that returns an error and checks that
WriteEvent returns an error, too.

antifuchs committed:
Add a test case for failing raw events, too

antifuchs committed:
Improve smartness of mock JSON endpoint

Now validates that the JSON data can be decoded, contains the
mandatory "event" key, and ensures that no keys other than the allowed
ones are present in the data.

antifuchs commented:
OK! I added two test cases to ensure that error responses get propagated, and made a JSON endpoint that validates the format of the JSON that gets posted (:

antifuchs committed:
Improve error HTTP response

fuyufjh commented:
Thanks for the awesome work! @antifuchs  I'll merge it soon
